crates/qpy/src/circuits.rs

```diff
@@ -0,0 +1,2174 @@
+                .get_bound(py)
+                .getattr(intern!(py, "__name__"))?,
+            StandardInstruction::Reset => imports::RESET
+                .get_bound(py)
+                .getattr(intern!(py, "__name__"))?,
+        }
+        .extract::<String>(),
+        OperationRef::Gate(pygate) => pygate
+            .gate
+            .bind(py)
+            .getattr(intern!(py, "__class__"))?
+            .getattr(intern!(py, "__name__"))?
+            .extract::<String>(),
+        OperationRef::Instruction(pyinst) => pyinst
+            .instruction
+            .bind(py)
+            .getattr(intern!(py, "__class__"))?
+            .getattr(intern!(py, "__name__"))?
+            .extract::<String>(),
+        OperationRef::Unitary(_) => Ok(UNITARY_GATE_CLASS_NAME.to_string()),
+        OperationRef::Operation(py_op) => py_op
+            .operation
+            .bind(py)
+            .getattr(intern!(py, "__class__"))?
+            .getattr(intern!(py, "__name__"))?
+            .extract::<String>(),
+    }?;
+    Ok(name)
+}
+
+fn get_num_ctrl_qubits(py: Python, op: &PackedOperation) -> PyResult<u32> {
+    match op.view() {
+        OperationRef::StandardGate(gate) => Ok(gate.num_ctrl_qubits()),
+        OperationRef::Gate(py_gate) => py_gate
+            .gate
+            .getattr(py, "num_ctrl_qubits")?
+            .extract::<u32>(py),
+        OperationRef::Instruction(py_inst) => py_inst
+            .instruction
+            .getattr(py, "num_ctrl_qubits")?
+            .extract::<u32>(py),
+        _ => Ok(0),
+    }
+}
+
+fn get_ctrl_state(py: Python, op: &PackedOperation, num_ctrl_qubits: u32) -> PyResult<u32> {
+    match op.view() {
+        OperationRef::Gate(py_gate) => py_gate.gate.getattr(py, "ctrl_state")?.extract::<u32>(py),
+        OperationRef::Instruction(py_inst) => py_inst
+            .instruction
+            .getattr(py, "ctrl_state")?
+            .extract::<u32>(py),
+        _ => Ok((1 << num_ctrl_qubits) - 1),
+    }
+}
+
+fn get_instruction_params(
+    py: Python,
+    instruction: &PackedInstruction,
+    qpy_data: &QPYWriteData,
+) -> PyResult<Vec<formats::PackedParam>> {
+    // The instruction params we store are about being able to reconstruct the objects; they don't
+    // necessarily need to match one-to-one to the `params` field.
+    if let OperationRef::Instruction(inst) = instruction.op.view() {
+        if inst
+            .instruction
+            .bind(py)
+            .is_instance(imports::CONTROL_FLOW_SWITCH_CASE_OP.get_bound(py))?
+        {
+            let op = inst.instruction.bind(py);
+            let target = op.getattr("target")?;
+            let cases = op.call_method0("cases_specifier")?;
+            let cases_tuple = imports::BUILTIN_TUPLE.get_bound(py).call1((cases,))?;
+            return Ok(vec![
+                pack_param(py, &target.extract::<Param>()?, qpy_data)?,
+                pack_param(py, &cases_tuple.extract::<Param>()?, qpy_data)?,
```

Comment:
We should be careful here since these aren't actually `Param`s. We're trying to get rid of `Param::Obj` (the control flow PR gets us most of the way there). I think they'd need to just be `PyObject` for now here.

Reply:
Ok, changed `pack_param` to handle `Bound<PyAny>` objects instead. I still need to handle `Param` objects [here](https://github.com/gadial/qiskit/blob/cc1f0fbd07cb6c38c123c078523fcf4636d18fc5/crates/qpy/src/circuits.rs#L302), though, so I renamed the previous `pack_param` to `pack_param_obj`.